velox/exec/prefixsort/PrefixSortEncoder.h

```diff
@@ -28,52 +28,177 @@ namespace facebook::velox::exec::prefixsort {
+/// For an unsigned integer '0x aa bb cc dd'
+///
+/// uint32_t a = 0xaabbccdd;
+/// char* buff = (char*)&a;
+/// for (int i = 0; i < 4; i++) {
+///   printf("%02hhx ", (int)buff[i]);
+/// }
+///
+///  the result is : dd cc bb aa, this is equivalent to: buff = [dd, cc, bb, aa]
+///  If we swap the buff, and then we can compare two buffs from the first byte
+///  to last byte, the compare result is equal to value-compare.
+///
+///  For any two unsigned integers, a < b <==> ~a > ~b so we invert bits when
+///  descending order.
 template <>
-FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encode(int64_t value, char* row) {
-  const auto v = __builtin_bswap64(static_cast<uint64_t>(value));
-  simd::memcpy(row, &v, sizeof(int64_t));
-  row[0] = flipSignBit(row[0]);
+FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encodeNoNulls(
+    uint32_t value,
+    char* dest) const {
+  auto& v = *reinterpret_cast<uint32_t*>(dest);
+  v = __builtin_bswap32(value);
+  if (!ascending_) {
+    v = ~v;
+  }
 }
 
+/// Compare two positive signed integers: storage layout is as same as unsigned
+/// integer, their sign-bit are same, flip sign-bit do not change result.
+/// Compare two negative signed integers: -n = ~n + 1, we can trait ~n + 1 as an
+/// unsigned integer, so the logic is as same as unsigned integer, also flip
+/// sign-bit do not change result.
+/// Compare positive vs negative: flip sign-bit to promise that positive
+/// always bigger than negative.
 template <>
-FOLLY_ALWAYS_INLINE void PrefixSortEncoder::decode<int64_t>(char* row) {
-  row[0] = flipSignBit(row[0]);
-  const auto v = __builtin_bswap64(*reinterpret_cast<uint64_t*>(row));
-  simd::memcpy(row, &v, sizeof(int64_t));
+FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encodeNoNulls(
+    int32_t value,
+    char* dest) const {
+  encodeNoNulls((uint32_t)(value ^ 1u << 31), dest);
 }
 
-/// For testing only.
-namespace {
-template <typename T>
-FOLLY_ALWAYS_INLINE void testingEncodeInPlace(const std::vector<T>& data) {
-  for (auto i = 0; i < data.size(); i++) {
-    PrefixSortEncoder::encode(data[i], (char*)data.data() + i * sizeof(T));
+/// Logic is nearly as same as int32_t
+template <>
+FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encodeNoNulls(
+    uint64_t value,
+    char* dest) const {
+  auto& v = *reinterpret_cast<uint64_t*>(dest);
+  v = __builtin_bswap64(value);
+  if (!ascending_) {
+    v = ~v;
   }
 }
 
-template <typename T>
-FOLLY_ALWAYS_INLINE void testingDecodeInPlace(const std::vector<T>& data) {
-  for (auto i = 0; i < data.size(); i++) {
-    PrefixSortEncoder::decode<T>((char*)data.data() + i * sizeof(T));
+template <>
+FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encodeNoNulls(
+    int64_t value,
+    char* dest) const {
+  encodeNoNulls((uint64_t)(value ^ 1ull << 63), dest);
+}
+
+/// Convert double to a uint64_t, their value comparison semantics remain
+/// consistent
+static FOLLY_ALWAYS_INLINE uint64_t encodeDouble(double x) {
+  uint64_t buff;
```

Comment:
Move this variable closer to where it is used. Wondering if there is a more descriptive name.